File: fixtures/nodejs/j05/fixture.js
// E-commerce API with 5 planted security vulnerabilities
// The model must find ALL of them

const express = require('express');
const jwt = require('jsonwebtoken');
const sqlite3 = require('better-sqlite3');

const app = express();
app.use(express.json());

const db = new sqlite3(':memory:');
db.exec(`CREATE TABLE users (id INTEGER PRIMARY KEY, email TEXT, role TEXT, password TEXT)`);
db.exec(`CREATE TABLE orders (id INTEGER PRIMARY KEY, user_id INTEGER, amount REAL)`);

// VULNERABILITY 1: Hardcoded API secret
const API_SECRET = 'super-secret-key-hardcoded-12345';

// Auth middleware
function authenticate(req, res, next) {
  const token = req.headers.authorization?.replace('Bearer ', '');
  if (!token) return res.status(401).json({ error: 'No token' });

  // VULNERABILITY 2: JWT decoded without verifying expiration
  const decoded = jwt.decode(token); // should be jwt.verify()
  if (!decoded) return res.status(401).json({ error: 'Invalid token' });
  req.user = decoded;
  next();
}

// Get user orders
app.get('/users/:id/orders', authenticate, (req, res) => {
  const userId = req.params.id;

  // VULNERABILITY 3: SQL injection via string concatenation
  const orders = db.prepare(`SELECT * FROM orders WHERE user_id = '${userId}'`).all();
  res.json(orders);
});

// Search users
app.get('/users/search', authenticate, (req, res) => {
  const { email } = req.query;
  const user = db.prepare("SELECT * FROM users WHERE email = ?").get(email);

  if (user) {
    // VULNERABILITY 4: PII logged in plaintext
    console.log(`User login: email=${user.email}, password=${user.password}, role=${user.role}`);
  }
  res.json(user || null);
});

// VULNERABILITY 5: Admin endpoint with no authentication check
app.delete('/admin/users/:id', (req, res) => {
  const { id } = req.params;
  db.prepare('DELETE FROM users WHERE id = ?').run(id);
  res.json({ deleted: id });
});

module.exports = { app };

File: fixtures/java/j05/Fixture.java
// E-commerce API with 5 planted security vulnerabilities
// The model must find ALL of them

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.Claims;

class ECommerceApi {

    private Connection db;

    // VULNERABILITY 1: Hardcoded API secret
    private static final String API_SECRET = "super-secret-key-hardcoded-12345";

    ECommerceApi() throws Exception {
        db = DriverManager.getConnection("jdbc:sqlite::memory:");
        Statement stmt = db.createStatement();
        stmt.execute("CREATE TABLE users (id INTEGER PRIMARY KEY, email TEXT, role TEXT, password TEXT)");
        stmt.execute("CREATE TABLE orders (id INTEGER PRIMARY KEY, user_id INTEGER, amount REAL)");
    }

    // Auth filter
    Claims authenticate(HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return null; // 401 No token
        }
        String token = authHeader.substring(7);

        // VULNERABILITY 2: JWT decoded without verifying signature or expiration
        Claims claims = Jwts.parserBuilder().build()
                .parseClaimsJwt(token) // should use parseClaimsJws() with signing key
                .getBody();
        return claims;
    }

    // Get user orders
    ResultSet getUserOrders(HttpServletRequest request, String userId) throws Exception {
        Claims user = authenticate(request);
        if (user == null) return null;

        // VULNERABILITY 3: SQL injection via string concatenation
        Statement stmt = db.createStatement();
        ResultSet rs = stmt.executeQuery("SELECT * FROM orders WHERE user_id = '" + userId + "'");
        return rs;
    }

    // Search users
    Object searchUser(HttpServletRequest request, String email) throws Exception {
        Claims user = authenticate(request);
        if (user == null) return null;

        PreparedStatement ps = db.prepareStatement("SELECT * FROM users WHERE email = ?");
        ps.setString(1, email);
        ResultSet rs = ps.executeQuery();

        if (rs.next()) {
            String foundEmail = rs.getString("email");
            String password = rs.getString("password");
            String role = rs.getString("role");
            // VULNERABILITY 4: PII logged in plaintext
            System.out.println("User login: email=" + foundEmail + ", password=" + password + ", role=" + role);
        }
        return rs;
    }

    // VULNERABILITY 5: Admin endpoint with no authentication check
    void deleteUser(String id) throws Exception {
        PreparedStatement ps = db.prepareStatement("DELETE FROM users WHERE id = ?");
        ps.setInt(1, Integer.parseInt(id));
        ps.executeUpdate();
        // returns { deleted: id } equivalent
    }
}
